--- a/drivers/bluetooth/btusb.c
+++ b/drivers/bluetooth/btusb.c
@@ -368,6 +368,10 @@
 
 	/* Additional Realtek 8723BU Bluetooth devices */
 	{ USB_DEVICE(0x7392, 0xa611), .driver_info = BTUSB_REALTEK },
+	
+	/* Tricolor 8723BU bt suport*/
+	{ USB_DEVICE(0x0bda, 0xb720), .driver_info = BTUSB_REALTEK },
+	{ USB_DEVICE(0x0bda, 0xb72a), .driver_info = BTUSB_REALTEK },
 
 	/* Additional Realtek 8723DE Bluetooth devices */
 	{ USB_DEVICE(0x0bda, 0xb009), .driver_info = BTUSB_REALTEK },
