--- ./arch/arm/tools/mach-types	2020-03-07 10:48:49.095980878 -0800
+++ /home/legend/Nuvoton/nuc980/kernel/NUC970-linux-4.4.194/arch/arm/tools/mach-types	2019-12-29 19:12:21.000000000 -0800
@@ -550,6 +550,7 @@
 u8520			MACH_U8520		U8520			3990
 urt			MACH_URT		URT			4347
 keystone		MACH_KEYSTONE		KEYSTONE		4390
+nuc970                  MACH_NUC970             NUC970                  4577
 ckb_rza1h		MACH_CKB_RZA1H		CKB_RZA1H		4780
 bcm2835			MACH_BCM2835		BCM2835			4828
 cm_3g			MACH_CM_3G		CM_3G			4943
