--- a/scripts/Makefile	2022-04-30 10:57:40.617382947 +0200
+++ b/scripts/Makefilew	2022-04-30 10:50:36.808295068 +0200
@@ -29,7 +29,7 @@
 HOSTCFLAGS_sign-file.o = $(CRYPTO_CFLAGS)
 HOSTLDLIBS_sign-file = $(CRYPTO_LIBS)
 HOSTCFLAGS_extract-cert.o = $(CRYPTO_CFLAGS)
-HOSTLDLIBS_extract-cert = $(CRYPTO_LIBS)
+HOSTLDLIBS_extract-cert = $(CRYPTO_LIBS) -lcrypto -lpthread
 
 always		:= $(hostprogs-y) $(hostprogs-m)
 
