--- a/arch/arm64/boot/dts/rockchip/rk3568-h68k-opc-board.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3568-h68k-opc-board.dts
@@ -146,7 +146,7 @@ regulator-state-mem {
 
 &gmac0 {
 	phy-mode = "rgmii";
-	clock_in_out = "input";
+	clock_in_out = "output";
 
 	snps,reset-gpio = <&gpio2 RK_PD3 GPIO_ACTIVE_LOW>;
 	snps,reset-active-low;
@@ -154,18 +154,18 @@ &gmac0 {
 	snps,reset-delays-us = <0 20000 100000>;
 
 	assigned-clocks = <&cru SCLK_GMAC0_RX_TX>, <&cru SCLK_GMAC0>;
-	assigned-clock-parents = <&cru SCLK_GMAC0_RGMII_SPEED>, <&gmac0_clkin>;
+	assigned-clock-parents = <&cru SCLK_GMAC0_RGMII_SPEED>, <&cru CLK_MAC0_2TOP>;
+	assigned-clock-rates = <0>, <125000000>;
 
 	pinctrl-names = "default";
 	pinctrl-0 = <&gmac0_miim
 		     &gmac0_tx_bus2
 		     &gmac0_rx_bus2
 		     &gmac0_rgmii_clk
-		     &gmac0_rgmii_bus
-		     &gmac0_clkinout>;
+		     &gmac0_rgmii_bus>;
 
-	tx_delay = <0x4f>;
-	rx_delay = <0x0f>;
+	tx_delay = <0x3c>;
+	rx_delay = <0x2f>;
 
 	phy-handle = <&rgmii_phy0>;
 	status = "okay";
@@ -173,7 +173,7 @@ &gmac0_rgmii_bus
 
 &gmac1 {
 	phy-mode = "rgmii";
-	clock_in_out = "input";
+	clock_in_out = "output";
 
 	snps,reset-gpio = <&gpio1 RK_PB0 GPIO_ACTIVE_LOW>;
 	snps,reset-active-low;
@@ -181,23 +181,22 @@ &gmac1 {
 	snps,reset-delays-us = <0 20000 100000>;
 
 	assigned-clocks = <&cru SCLK_GMAC1_RX_TX>, <&cru SCLK_GMAC1>;
-	assigned-clock-parents = <&cru SCLK_GMAC1_RGMII_SPEED>, <&gmac1_clkin>;
+	assigned-clock-parents = <&cru SCLK_GMAC1_RGMII_SPEED>, <&cru CLK_MAC1_2TOP>;
+	assigned-clock-rates = <0>, <125000000>;
 
 	pinctrl-names = "default";
 	pinctrl-0 = <&gmac1m1_miim
 		     &gmac1m1_tx_bus2
 		     &gmac1m1_rx_bus2
 		     &gmac1m1_rgmii_clk
-		     &gmac1m1_rgmii_bus
-		     &gmac1m1_clkinout>;
+		     &gmac1m1_rgmii_bus>;
 
-	tx_delay = <0x3c>;
-	rx_delay = <0x2f>;
+	tx_delay = <0x4f>;
+	rx_delay = <0x26>;
 
 	phy-handle = <&rgmii_phy1>;
 	status = "okay";
 };
-
 &i2c0 {
 	status = "okay";
 
-- 
2.34.1

